src/server/frontend_wayland/xdg_activation_v1.cpp

```diff
@@ -0,0 +1,312 @@
+    class Instance : public wayland::XdgActivationV1
+    {
+    public:
+        Instance(
+            struct wl_resource* resource,
+            mf::XdgActivationV1* xdg_activation_v1,
+            std::shared_ptr<msh::Shell> const& shell,
+            std::shared_ptr<msh::FocusController> const& focus_controller,
+            std::shared_ptr<MainLoop> const& main_loop);
+
+    private:
+        void get_activation_token(struct wl_resource* id) override;
+        void activate(std::string const& token, struct wl_resource* surface) override;
+
+        mf::XdgActivationV1* xdg_activation_v1;
+        std::shared_ptr<msh::Shell> shell;
+        std::shared_ptr<msh::FocusController> const focus_controller;
+        std::shared_ptr<MainLoop> main_loop;
+    };
+
+    struct Token
+    {
+        std::string token;
+        std::unique_ptr<time::Alarm> alarm;
+    };
+
+    void bind(wl_resource* resource) override;
+
+    std::shared_ptr<msh::Shell> shell;
+    std::shared_ptr<msh::FocusController> const focus_controller;
+    std::shared_ptr<MainLoop> main_loop;
+    std::vector<Token> pending_tokens;
+    std::mutex lock;
+};
+
+class XdgActivationTokenV1 : public wayland::XdgActivationTokenV1
+{
+public:
+    XdgActivationTokenV1(
+        struct wl_resource* resource,
+        std::string const& token,
+        std::shared_ptr<msh::FocusController> const& focus_controller);
+
+private:
+    void set_serial(uint32_t serial, struct wl_resource* seat) override;
+    void set_app_id(std::string const& app_id) override;
+    void set_surface(struct wl_resource* surface) override;
+    void commit() override;
+
+    std::string token;
+    std::shared_ptr<msh::FocusController> const focus_controller;
+    std::optional<uint32_t> serial;
+    std::optional<struct wl_resource*> seat;
+    std::optional<std::string> app_id;
+    std::optional<struct wl_resource*> requesting_surface;
+};
+}
+
+auto mf::create_xdg_activation_v1(
+    struct wl_display* display,
+    std::shared_ptr<msh::Shell> const& shell,
+    std::shared_ptr<msh::FocusController> const& focus_controller,
+    std::shared_ptr<MainLoop> const& main_loop)
+    -> std::shared_ptr<mw::XdgActivationV1::Global>
+{
+    return std::make_shared<mf::XdgActivationV1>(display, shell, focus_controller, main_loop);
+}
+
+mf::XdgActivationV1::XdgActivationV1(
+    struct wl_display* display,
+    std::shared_ptr<msh::Shell> const& shell,
+    std::shared_ptr<msh::FocusController> const& focus_controller,
+    std::shared_ptr<MainLoop> const& main_loop)
+    : Global(display, Version<1>()),
+      shell{shell},
+      focus_controller{focus_controller},
+      main_loop{main_loop}
+{
+
+}
+
+std::string const& mf::XdgActivationV1::create_token()
+{
+    auto generated = generate_token();
+    Token token{generated, main_loop->create_alarm([this, generated]()
+    {
+        try_consume_token(generated);
+    })};
+    token.alarm->reschedule_in(TIMEOUT_MS);
+
+    std::lock_guard guard(lock);
```

Comment:
It is strange seeing a lock here - shouldn't all this code run on the Wayland thread?

Reply:
The lock is because we're also setting an `Alarm` for timeout on the token. That would be on the main thread, no?